sbt-scalajs-bundler/src/main/scala/scalajsbundler/util/JS.scala

```diff
@@ -6,39 +6,53 @@ import org.scalajs.core.ir.Position
 import org.scalajs.core.tools.javascript.Printers
 import org.scalajs.core.tools.javascript.Trees._
 
+private[util] sealed abstract class JSLike(val tree: Tree) {
+  def show: String = tree.show
+  def show(isStat: Boolean = true): String = JSLike.show(tree, isStat)
+  def toJson: String = JSLike.show(tree, isStat = false)
+}
+
+object JSLike {
+  def show(tree: Tree, isStat: Boolean = true): String = {
+    val writer = new java.io.StringWriter
+    val printer = new Printers.JSTreePrinter(writer)
+    printer.printTree(tree, isStat)
+    writer.toString
+  }
+}
+
 /** A convenient wrapper around JS trees */
-final case class JS(tree: Tree) extends AnyVal {
+final class JS private(tree: Tree) extends JSLike(tree) {
   import JS.position
   def dot(ident: String): JS = JS(DotSelect(tree, Ident(ident)))
   def bracket(ident: String): JS = JS(BracketSelect(tree, StringLiteral(ident)))
-  def bracket(ident: JS): JS = JS(BracketSelect(tree, ident.tree))
-  def assign(rhs: JS): JS = JS(Assign(tree, rhs.tree))
-  def apply(args: JS*): JS = JS(Apply(tree, args.map(_.tree).to[List]))
-
-  def show: String = tree.show
+  def bracket(ident: JSLike): JS = JS(BracketSelect(tree, ident.tree))
+  def assign(rhs: JSLike): JS = JS(Assign(tree, rhs.tree))
+  def apply(args: JSLike*): JS = JS(Apply(tree, args.map(_.tree).to[List]))
 }
 
 object JS {
 
   implicit lazy val position: Position = Position.NoPosition
 
-  /** String literal */
-  def str(value: String): JS =
-    JS(StringLiteral(value))
+  def apply(tree: Tree): JS = new JS(tree)
+
+  /** Array literal. */
+  def arr(elems: JSLike*): JS = JS(ArrayConstr(elems.map(_.tree).to[List]))
 
   /** Boolean literal */
   def bool(value: Boolean): JS = JS(BooleanLiteral(value))
 
   /** Object literal */
-  def obj(fields: (String, JS)*): JS =
+  def obj(fields: (String, JSLike)*): JS =
     JS(ObjectConstr(fields.map { case (ident, value) => (StringLiteral(ident), value.tree) }.to[List]))
```

Comment:
It's not quite duplicated though. The JSON definition takes `(String, JSON)*` to explicitly disallow any `JS` stuff to be mixed in. Removing `JS.obj` will make [this](https://github.com/scalacenter/scalajs-bundler/blob/fb8064bc888f2c829d86daaea6e89144c507d8fe/sbt-scalajs-bundler/src/main/scala/scalajsbundler/Webpack.scala#L76) invalid since it's adding a `RegExp` type.